app/src/main/kotlin/com/wire/android/ui/home/messagecomposer/attachment/AttachmentOptions.kt

```diff
@@ -9,27 +9,50 @@ import androidx.compose.foundation.layout.PaddingValues
 import androidx.compose.foundation.layout.fillMaxWidth
 import androidx.compose.foundation.lazy.GridCells
 import androidx.compose.foundation.lazy.LazyVerticalGrid
+import androidx.compose.material3.ExperimentalMaterial3Api
+import androidx.compose.material3.SnackbarHostState
 import androidx.compose.runtime.Composable
+import androidx.compose.runtime.LaunchedEffect
 import androidx.compose.ui.ExperimentalComposeUiApi
 import androidx.compose.ui.Modifier
+import androidx.compose.ui.platform.LocalContext
 import androidx.compose.ui.res.stringResource
 import androidx.compose.ui.tooling.preview.Preview
+import androidx.hilt.navigation.compose.hiltViewModel
 import com.wire.android.R
 import com.wire.android.appLogger
 import com.wire.android.ui.common.AttachmentButton
 import com.wire.android.ui.common.dimensions
+import com.wire.android.ui.home.conversations.AttachmentPart
 import com.wire.android.util.permission.UseCameraRequestFlow
+import com.wire.android.util.permission.UseStorageRequestFlow
 import com.wire.android.util.permission.rememberCaptureVideoFlow
 import com.wire.android.util.permission.rememberCurrentLocationFlow
 import com.wire.android.util.permission.rememberOpenFileBrowserFlow
 import com.wire.android.util.permission.rememberOpenGalleryFlow
 import com.wire.android.util.permission.rememberRecordAudioRequestFlow
 import com.wire.android.util.permission.rememberTakePictureFlow
 
-@OptIn(ExperimentalComposeUiApi::class, ExperimentalFoundationApi::class)
+@OptIn(ExperimentalComposeUiApi::class, ExperimentalFoundationApi::class, ExperimentalMaterial3Api::class)
 @Composable
-fun AttachmentOptionsComponent() {
-    val attachmentOptions = buildAttachmentOptionItems()
+fun AttachmentOptionsComponent(onSendAttachment: (AttachmentPart?) -> Unit, snackbarHostState: SnackbarHostState) {
+    val viewModel: AttachmentOptionsViewModel = hiltViewModel()
```

Comment:
from what I see this viewmodel is handeling the UI logic only, the problem that I see with using a hiltViewModel inside a @Composable is that theoritcally when we would have this function called two times inside a another @Composable, so lets say some screen would use this AttachmentOptionsComponent in two different places, those two instances of this composable will share common state as the hiltViewModel() uses a NavBackStackEntry as a host for the viewmodel, i know that for our case this will probably not happen :P 

another problem I see with this approach is that lets say we show the AttachmentOptionsComponent() we fire prepareAttachment() but then user closes the AttachmentOptionsComponent()  but before closing fires this coroutine , then it will not be cancelled, I do not know if that is something we want ? 
```

   viewModelScope.launch {
            attachmentState = try {
                val attachment =
                    AttachmentPart(attachmentUri.getMimeType(context).orDefault(DEFAULT_FILE_MIME_TYPE), attachmentUri.toByteArray(context))
                AttachmentState.Picked(attachment)
            } catch (e: IOException) {
                AttachmentState.Error
            }
        }
```

if you refactored the code so that is uses a pure kotlin Class like MessageComposer for example and uses a remember { } retuning that pure Kotlin Class, you will have a proper coroutine cancelation and also each call to AttachmentOptionsComponent will create its own state object managing its own state and do not share it with other calls to AttachmentOptionsComponent() I even wonder if it makes sense to create a seperate plain Kotlin Class to manage this UI component , since this is going to be used only inside the MessageComposer ? So I think it even makes sense to move all this logic to MessageComposerInnerState class, let me know what you tihnk of it

Reply:
Wow! long but insightful read, thanks 👏  I just want to point out that this ViewModel itself handle the state of the view, but the main idea behind the creation was using a ViewModelScope coroutine to do the "heavy" IO work, cause transforming a `Uri` to a `ByteArray` fits better outside UI context.

Having said that, I agree with the point that actually this will fit better inside the `MessageComposerInnerState` class, cause in some degree this it's also part of composing a message :D 

I will try that approach, fitting this Attachment States inside `MessageComposerInnerState` and maybe just mark the `prepareAttachment` func as suspend to Launch it from the Composable component.

Reply:
@trOnk12 already implemented the suggested solution I think now looks better in fact without the ViewModel 🚀